monitoring-plugins fails to build with OpenSSL 4.0

## monitoring-plugins Version

monitoring-plugins https://github.com/monitoring-plugins/monitoring-plugins/commit/b6021e5e372ba9219684a809e73a20f50e50caca

## System information
Arch Linux Rolling

## Describe the bug
monitoring-plugins fails to build against OpenSSL 4.0 as the `ASN1_STRING` structure has been made opaque and now requires use via access functions.

## Additional information
Steps to reproduce:
```
$ git clone https://github.com/monitoring-plugins/monitoring-plugins.git
$ cd monitoring-plugins
$ autoreconf -fiv
$ ./configure
$ make
....
sslutils.c: In function ‘np_net_ssl_check_certificate’:
sslutils.c:202:27: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  202 |         X509_NAME *subj = X509_get_subject_name(certificate);
      |                           ^~~~~~~~~~~~~~~~~~~~~
sslutils.c:210:9: warning: ‘X509_NAME_get_text_by_NID’ is deprecated: Since OpenSSL 4.0 [-Wdeprecated-declarations]
  210 |         int cnlen = X509_NAME_get_text_by_NID(subj, NID_commonName, cn, sizeof(cn));
      |         ^~~
In file included from common.h:152,
                 from sslutils.c:31:
/usr/include/openssl/x509.h:1041:27: note: declared here
 1041 | OSSL_DEPRECATEDIN_4_0 int X509_NAME_get_text_by_NID(const X509_NAME *name,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~
sslutils.c:221:15: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  221 |         if (tm->type == V_ASN1_UTCTIME) {
      |               ^~
sslutils.c:222:23: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  222 |                 if (tm->length < 10) {
      |                       ^~
sslutils.c:226:36: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  226 |                 stamp.tm_year = (tm->data[0] - '0') * 10 + (tm->data[1] - '0');
      |                                    ^~
sslutils.c:226:63: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  226 |                 stamp.tm_year = (tm->data[0] - '0') * 10 + (tm->data[1] - '0');
      |                                                               ^~
sslutils.c:233:23: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  233 |                 if (tm->length < 12) {
      |                       ^~
sslutils.c:237:36: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  237 |                 stamp.tm_year = (tm->data[0] - '0') * 1000 + (tm->data[1] - '0') * 100 +
      |                                    ^~
sslutils.c:237:65: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  237 |                 stamp.tm_year = (tm->data[0] - '0') * 1000 + (tm->data[1] - '0') * 100 +
      |                                                                 ^~
sslutils.c:238:52: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  238 |                                                 (tm->data[2] - '0') * 10 + (tm->data[3] - '0');
      |                                                    ^~
sslutils.c:238:79: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  238 |                                                 (tm->data[2] - '0') * 10 + (tm->data[3] - '0');
      |                                                                               ^~
sslutils.c:242:27: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  242 |         stamp.tm_mon = (tm->data[2 + offset] - '0') * 10 + (tm->data[3 + offset] - '0') - 1;
      |                           ^~
sslutils.c:242:63: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  242 |         stamp.tm_mon = (tm->data[2 + offset] - '0') * 10 + (tm->data[3 + offset] - '0') - 1;
      |                                                               ^~
sslutils.c:243:28: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  243 |         stamp.tm_mday = (tm->data[4 + offset] - '0') * 10 + (tm->data[5 + offset] - '0');
      |                            ^~
sslutils.c:243:64: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  243 |         stamp.tm_mday = (tm->data[4 + offset] - '0') * 10 + (tm->data[5 + offset] - '0');
      |                                                                ^~
sslutils.c:244:28: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  244 |         stamp.tm_hour = (tm->data[6 + offset] - '0') * 10 + (tm->data[7 + offset] - '0');
      |                            ^~
sslutils.c:244:64: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  244 |         stamp.tm_hour = (tm->data[6 + offset] - '0') * 10 + (tm->data[7 + offset] - '0');
      |                                                                ^~
sslutils.c:245:27: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  245 |         stamp.tm_min = (tm->data[8 + offset] - '0') * 10 + (tm->data[9 + offset] - '0');
      |                           ^~
sslutils.c:245:63: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  245 |         stamp.tm_min = (tm->data[8 + offset] - '0') * 10 + (tm->data[9 + offset] - '0');
      |                                                               ^~
sslutils.c:246:27: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  246 |         stamp.tm_sec = (tm->data[10 + offset] - '0') * 10 + (tm->data[11 + offset] - '0');
      |                           ^~
sslutils.c:246:64: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  246 |         stamp.tm_sec = (tm->data[10 + offset] - '0') * 10 + (tm->data[11 + offset] - '0');
      |                                                                ^~
sslutils.c: In function ‘np_net_ssl_get_cert_expiration’:
sslutils.c:329:27: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  329 |         X509_NAME *subj = X509_get_subject_name(certificate);
      |                           ^~~~~~~~~~~~~~~~~~~~~
sslutils.c:338:9: warning: ‘X509_NAME_get_text_by_NID’ is deprecated: Since OpenSSL 4.0 [-Wdeprecated-declarations]
  338 |         int cnlen = X509_NAME_get_text_by_NID(subj, NID_commonName, cn, sizeof(cn));
      |         ^~~
/usr/include/openssl/x509.h:1041:27: note: declared here
 1041 | OSSL_DEPRECATEDIN_4_0 int X509_NAME_get_text_by_NID(const X509_NAME *name,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~
sslutils.c:349:33: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  349 |         if (expiration_timestamp->type == V_ASN1_UTCTIME) {
      |                                 ^~
sslutils.c:350:41: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  350 |                 if (expiration_timestamp->length < 10) {
      |                                         ^~
sslutils.c:356:46: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  356 |                         (expiration_timestamp->data[0] - '0') * 10 + (expiration_timestamp->data[1] - '0');
      |                                              ^~
sslutils.c:356:91: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  356 |                         (expiration_timestamp->data[0] - '0') * 10 + (expiration_timestamp->data[1] - '0');
      |                                                                                           ^~
sslutils.c:362:41: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  362 |                 if (expiration_timestamp->length < 12) {
      |                                         ^~
sslutils.c:367:54: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  367 |                 stamp.tm_year = (expiration_timestamp->data[0] - '0') * 1000 +
      |                                                      ^~
sslutils.c:368:70: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  368 |                                                 (expiration_timestamp->data[1] - '0') * 100 +
      |                                                                      ^~
sslutils.c:369:70: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  369 |                                                 (expiration_timestamp->data[2] - '0') * 10 +
      |                                                                      ^~
sslutils.c:370:70: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  370 |                                                 (expiration_timestamp->data[3] - '0');
      |                                                                      ^~
sslutils.c:374:45: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  374 |         stamp.tm_mon = (expiration_timestamp->data[2 + offset] - '0') * 10 +
      |                                             ^~
sslutils.c:375:57: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  375 |                                    (expiration_timestamp->data[3 + offset] - '0') - 1;
      |                                                         ^~
sslutils.c:376:46: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  376 |         stamp.tm_mday = (expiration_timestamp->data[4 + offset] - '0') * 10 +
      |                                              ^~
sslutils.c:377:62: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  377 |                                         (expiration_timestamp->data[5 + offset] - '0');
      |                                                              ^~
sslutils.c:378:46: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  378 |         stamp.tm_hour = (expiration_timestamp->data[6 + offset] - '0') * 10 +
      |                                              ^~
sslutils.c:379:62: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  379 |                                         (expiration_timestamp->data[7 + offset] - '0');
      |                                                              ^~
sslutils.c:380:45: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  380 |         stamp.tm_min = (expiration_timestamp->data[8 + offset] - '0') * 10 +
      |                                             ^~
sslutils.c:381:57: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  381 |                                    (expiration_timestamp->data[9 + offset] - '0');
      |                                                         ^~
sslutils.c:382:45: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  382 |         stamp.tm_sec = (expiration_timestamp->data[10 + offset] - '0') * 10 +
      |                                             ^~
sslutils.c:383:57: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  383 |                                    (expiration_timestamp->data[11 + offset] - '0');
      |                                                         ^~
sslutils.c: In function ‘mp_net_ssl_check_certificate’:
sslutils.c:470:27: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  470 |         X509_NAME *subj = X509_get_subject_name(certificate);
      |                           ^~~~~~~~~~~~~~~~~~~~~
sslutils.c:479:9: warning: ‘X509_NAME_get_text_by_NID’ is deprecated: Since OpenSSL 4.0 [-Wdeprecated-declarations]
  479 |         int cnlen = X509_NAME_get_text_by_NID(subj, NID_commonName, commonName, sizeof(commonName));
      |         ^~~
/usr/include/openssl/x509.h:1041:27: note: declared here
 1041 | OSSL_DEPRECATEDIN_4_0 int X509_NAME_get_text_by_NID(const X509_NAME *name,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~
sslutils.c:490:29: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  490 |         if (expiry_timestamp->type == V_ASN1_UTCTIME) {
      |                             ^~
sslutils.c:491:37: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  491 |                 if (expiry_timestamp->length < 10) {
      |                                     ^~
sslutils.c:497:50: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  497 |                 stamp.tm_year = (expiry_timestamp->data[0] - '0') * 10 + (expiry_timestamp->data[1] - '0');
      |                                                  ^~
sslutils.c:497:91: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  497 |                 stamp.tm_year = (expiry_timestamp->data[0] - '0') * 10 + (expiry_timestamp->data[1] - '0');
      |                                                                                           ^~
sslutils.c:504:37: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  504 |                 if (expiry_timestamp->length < 12) {
      |                                     ^~
sslutils.c:509:50: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  509 |                 stamp.tm_year = (expiry_timestamp->data[0] - '0') * 1000 +
      |                                                  ^~
sslutils.c:510:66: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  510 |                                                 (expiry_timestamp->data[1] - '0') * 100 +
      |                                                                  ^~
sslutils.c:511:66: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  511 |                                                 (expiry_timestamp->data[2] - '0') * 10 + (expiry_timestamp->data[3] - '0');
      |                                                                  ^~
sslutils.c:511:107: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  511 |                                                 (expiry_timestamp->data[2] - '0') * 10 + (expiry_timestamp->data[3] - '0');
      |                                                                                                           ^~
sslutils.c:516:41: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  516 |         stamp.tm_mon = (expiry_timestamp->data[2 + offset] - '0') * 10 +
      |                                         ^~
sslutils.c:517:53: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  517 |                                    (expiry_timestamp->data[3 + offset] - '0') - 1;
      |                                                     ^~
sslutils.c:518:42: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  518 |         stamp.tm_mday = (expiry_timestamp->data[4 + offset] - '0') * 10 +
      |                                          ^~
sslutils.c:519:58: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  519 |                                         (expiry_timestamp->data[5 + offset] - '0');
      |                                                          ^~
sslutils.c:520:42: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  520 |         stamp.tm_hour = (expiry_timestamp->data[6 + offset] - '0') * 10 +
      |                                          ^~
sslutils.c:521:58: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  521 |                                         (expiry_timestamp->data[7 + offset] - '0');
      |                                                          ^~
sslutils.c:522:41: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  522 |         stamp.tm_min = (expiry_timestamp->data[8 + offset] - '0') * 10 +
      |                                         ^~
sslutils.c:523:53: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  523 |                                    (expiry_timestamp->data[9 + offset] - '0');
      |                                                     ^~
sslutils.c:524:41: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  524 |         stamp.tm_sec = (expiry_timestamp->data[10 + offset] - '0') * 10 +
      |                                         ^~
sslutils.c:525:53: error: invalid use of incomplete typedef ‘ASN1_STRING’ {aka ‘struct asn1_string_st’}
  525 |                                    (expiry_timestamp->data[11 + offset] - '0');
      |                                                     ^~
....
```
[monitoring-plugins-openssl-4.0-build-failure.log](https://github.com/user-attachments/files/30368635/monitoring-plugins-openssl-4.0-build-failure.log)

@pierres

--- a/.gitignore
+++ b/.gitignore
@@ -231,6 +231,7 @@ plugins/check_disk.d/.dirstamp
 # /plugins/tests/
 /plugins/tests/Makefile
 /plugins/tests/Makefile.in
+/plugins/tests/test_sslutils
 /plugins/tests/test_utils
 /plugins/tests/test_check_disk
 /plugins/tests/test_check_swap

--- a/configure.ac
+++ b/configure.ac
@@ -175,7 +175,7 @@ if test "$enable_libtap" = "yes" ; then
 	EXTRA_TEST="test_utils test_tcp test_cmd test_base64 test_generic_output"
 	AC_SUBST(EXTRA_TEST)
 
-	EXTRA_PLUGIN_TESTS="tests/test_check_swap tests/test_check_disk"
+	EXTRA_PLUGIN_TESTS="tests/test_check_swap tests/test_check_disk tests/test_sslutils"
 	AC_SUBST(EXTRA_PLUGIN_TESTS)
 fi
 

--- a/plugins/Makefile.am
+++ b/plugins/Makefile.am
@@ -49,13 +49,15 @@ EXTRA_PROGRAMS = check_mysql check_radius check_pgsql check_hpjd \
 	\
 	tests/test_check_swap \
 	tests/test_check_snmp \
-	tests/test_check_disk
+	tests/test_check_disk \
+	tests/test_sslutils
 
 SUBDIRS = picohttpparser
 
 np_test_scripts = tests/test_check_swap.t \
 				  tests/test_check_snmp.t \
-				  tests/test_check_disk.t
+				  tests/test_check_disk.t \
+				  tests/test_sslutils.t
 
 EXTRA_DIST = t \
 			 tests \
@@ -192,6 +194,8 @@ tests_test_check_snmp_LDADD = $(BASEOBJS) $(tap_ldflags) -ltap
 tests_test_check_snmp_SOURCES = tests/test_check_snmp.c check_snmp.d/check_snmp_helpers.c
 tests_test_check_disk_LDADD = $(BASEOBJS) $(tap_ldflags) check_disk.d/utils_disk.c -ltap
 tests_test_check_disk_SOURCES = tests/test_check_disk.c
+tests_test_sslutils_LDADD = $(SSLOBJS) $(tap_ldflags) -ltap
+tests_test_sslutils_SOURCES = tests/test_sslutils.c
 
 ##############################################################################
 # secondary dependencies

--- a/plugins/sslutils.c
+++ b/plugins/sslutils.c
@@ -47,6 +47,33 @@ int np_net_ssl_init_with_hostname_and_version(int sd, char *host_name, int versi
 	return np_net_ssl_init_with_hostname_version_and_cert(sd, host_name, version, NULL, NULL);
 }
 
+#	ifdef MOPL_USE_OPENSSL
+int np_net_asn1_time_to_time_t(const ASN1_TIME *asn1_time, time_t *out) {
+	struct tm tm = {};
+	if (!ASN1_TIME_to_tm(asn1_time, &tm)) {
+		return 0;
+	}
+	*out = timegm(&tm);
+	if (*out == (time_t)-1) {
+		return 0;
+	}
+	return 1;
+}
+
+void np_net_format_timestamp(time_t t, char *buf, size_t buflen) {
+	char *tz = getenv("TZ");
+	setenv("TZ", "GMT", 1);
+	tzset();
+	strftime(buf, buflen, "%c %z", localtime(&t));
+	if (tz) {
+		setenv("TZ", tz, 1);
+	} else {
+		unsetenv("TZ");
+	}
+	tzset();
+}
+#endif /* MOPL_USE_OPENSSL */
+
 int np_net_ssl_init_with_hostname_version_and_cert(int sd, char *host_name, int version, char *cert,
 												   char *privkey) {
 	long options = 0;

--- a/plugins/sslutils.c
+++ b/plugins/sslutils.c
@@ -240,55 +240,16 @@ mp_state_enum np_net_ssl_check_certificate(X509 *certificate, int days_till_exp_
 	}
 
 	/* Retrieve timestamp of certificate */
-	ASN1_STRING *tm = X509_get_notAfter(certificate);
-
-	int offset = 0;
-	struct tm stamp = {};
-	/* Generate tm structure to process timestamp */
-	if (tm->type == V_ASN1_UTCTIME) {
-		if (tm->length < 10) {
-			printf("%s\n", _("CRITICAL - Wrong time format in certificate."));
-			return STATE_CRITICAL;
-		}
-		stamp.tm_year = (tm->data[0] - '0') * 10 + (tm->data[1] - '0');
-		if (stamp.tm_year < 50) {
-			stamp.tm_year += 100;
-		}
-		offset = 0;
-
-	} else {
-		if (tm->length < 12) {
-			printf("%s\n", _("CRITICAL - Wrong time format in certificate."));
-			return STATE_CRITICAL;
-		}
-		stamp.tm_year = (tm->data[0] - '0') * 1000 + (tm->data[1] - '0') * 100 +
-						(tm->data[2] - '0') * 10 + (tm->data[3] - '0');
-		stamp.tm_year -= 1900;
-		offset = 2;
+	const ASN1_TIME *asn1_not_after = X509_get_notAfter(certificate);
+	time_t expiry_time;
+	if (!np_net_asn1_time_to_time_t(asn1_not_after, &expiry_time)) {
+		printf("%s\n", _("CRITICAL - Wrong time format in certificate."));
+		return STATE_CRITICAL;
 	}
-	stamp.tm_mon = (tm->data[2 + offset] - '0') * 10 + (tm->data[3 + offset] - '0') - 1;
-	stamp.tm_mday = (tm->data[4 + offset] - '0') * 10 + (tm->data[5 + offset] - '0');
-	stamp.tm_hour = (tm->data[6 + offset] - '0') * 10 + (tm->data[7 + offset] - '0');
-	stamp.tm_min = (tm->data[8 + offset] - '0') * 10 + (tm->data[9 + offset] - '0');
-	stamp.tm_sec = (tm->data[10 + offset] - '0') * 10 + (tm->data[11 + offset] - '0');
-	stamp.tm_isdst = -1;
-
-	time_t tm_t = timegm(&stamp);
-	float time_left = difftime(tm_t, time(NULL));
+	float time_left = difftime(expiry_time, time(NULL));
 	int days_left = time_left / 86400;
-	char *tz = getenv("TZ");
-	setenv("TZ", "GMT", 1);
-	tzset();
-
 	char timestamp[50] = "";
-	strftime(timestamp, 50, "%c %z", localtime(&tm_t));
-	if (tz) {
-		setenv("TZ", tz, 1);
-	} else {
-		unsetenv("TZ");
-	}
-
-	tzset();
+	np_net_format_timestamp(expiry_time, timestamp, sizeof(timestamp));
 
 	int time_remaining;
 	mp_state_enum status = STATE_UNKNOWN;
@@ -368,65 +329,14 @@ retrieve_expiration_time_result np_net_ssl_get_cert_expiration(X509 *certificate
 	}
 
 	/* Retrieve timestamp of certificate */
-	ASN1_STRING *expiration_timestamp = X509_get_notAfter(certificate);
-
-	int offset = 0;
-	struct tm stamp = {};
-	/* Generate tm structure to process timestamp */
-	if (expiration_timestamp->type == V_ASN1_UTCTIME) {
-		if (expiration_timestamp->length < 10) {
-			result.errors = WRONG_TIME_FORMAT_IN_CERTIFICATE;
-			return result;
-		}
-
-		stamp.tm_year =
-			(expiration_timestamp->data[0] - '0') * 10 + (expiration_timestamp->data[1] - '0');
-		if (stamp.tm_year < 50) {
-			stamp.tm_year += 100;
-		}
-		offset = 0;
-	} else {
-		if (expiration_timestamp->length < 12) {
-			result.errors = WRONG_TIME_FORMAT_IN_CERTIFICATE;
-			return result;
-		}
-
-		stamp.tm_year = (expiration_timestamp->data[0] - '0') * 1000 +
-						(expiration_timestamp->data[1] - '0') * 100 +
-						(expiration_timestamp->data[2] - '0') * 10 +
-						(expiration_timestamp->data[3] - '0');
-		stamp.tm_year -= 1900;
-		offset = 2;
+	const ASN1_TIME *asn1_not_after = X509_get_notAfter(certificate);
+	time_t expiry_time;
+	if (!np_net_asn1_time_to_time_t(asn1_not_after, &expiry_time)) {
+		result.errors = WRONG_TIME_FORMAT_IN_CERTIFICATE;
+		return result;
 	}
-	stamp.tm_mon = (expiration_timestamp->data[2 + offset] - '0') * 10 +
-				   (expiration_timestamp->data[3 + offset] - '0') - 1;
-	stamp.tm_mday = (expiration_timestamp->data[4 + offset] - '0') * 10 +
-					(expiration_timestamp->data[5 + offset] - '0');
-	stamp.tm_hour = (expiration_timestamp->data[6 + offset] - '0') * 10 +
-					(expiration_timestamp->data[7 + offset] - '0');
-	stamp.tm_min = (expiration_timestamp->data[8 + offset] - '0') * 10 +
-				   (expiration_timestamp->data[9 + offset] - '0');
-	stamp.tm_sec = (expiration_timestamp->data[10 + offset] - '0') * 10 +
-				   (expiration_timestamp->data[11 + offset] - '0');
-	stamp.tm_isdst = -1;
-
-	time_t tm_t = timegm(&stamp);
-	double time_left = difftime(tm_t, time(NULL));
+	double time_left = difftime(expiry_time, time(NULL));
 	result.remaining_seconds = time_left;
-
-	char *timezone = getenv("TZ");
-	setenv("TZ", "GMT", 1);
-	tzset();
-
-	char timestamp[50] = "";
-	strftime(timestamp, 50, "%c %z", localtime(&tm_t));
-	if (timezone) {
-		setenv("TZ", timezone, 1);
-	} else {
-		unsetenv("TZ");
-	}
-
-	tzset();
 
 	X509_free(certificate);
 
@@ -509,65 +419,17 @@ mp_subcheck mp_net_ssl_check_certificate(X509 *certificate, int days_till_exp_wa
 	}
 
 	/* Retrieve timestamp of certificate */
-	ASN1_STRING *expiry_timestamp = X509_get_notAfter(certificate);
-
-	int offset = 0;
-	struct tm stamp = {};
-	/* Generate tm structure to process timestamp */
-	if (expiry_timestamp->type == V_ASN1_UTCTIME) {
-		if (expiry_timestamp->length < 10) {
-			xasprintf(&sc_cert.output, _("Wrong time format in certificate"));
-			sc_cert = mp_set_subcheck_state(sc_cert, STATE_CRITICAL);
-			return sc_cert;
-		}
-
-		stamp.tm_year = (expiry_timestamp->data[0] - '0') * 10 + (expiry_timestamp->data[1] - '0');
-		if (stamp.tm_year < 50) {
-			stamp.tm_year += 100;
-		}
-
-		offset = 0;
-	} else {
-		if (expiry_timestamp->length < 12) {
-			xasprintf(&sc_cert.output, _("Wrong time format in certificate"));
-			sc_cert = mp_set_subcheck_state(sc_cert, STATE_CRITICAL);
-			return sc_cert;
-		}
-		stamp.tm_year = (expiry_timestamp->data[0] - '0') * 1000 +
-						(expiry_timestamp->data[1] - '0') * 100 +
-						(expiry_timestamp->data[2] - '0') * 10 + (expiry_timestamp->data[3] - '0');
-		stamp.tm_year -= 1900;
-		offset = 2;
+	const ASN1_TIME *asn1_not_after = X509_get_notAfter(certificate);
+	time_t expiry_time;
+	if (!np_net_asn1_time_to_time_t(asn1_not_after, &expiry_time)) {
+		xasprintf(&sc_cert.output, _("Wrong time format in certificate"));
+		sc_cert = mp_set_subcheck_state(sc_cert, STATE_CRITICAL);
+		return sc_cert;
 	}
-
-	stamp.tm_mon = (expiry_timestamp->data[2 + offset] - '0') * 10 +
-				   (expiry_timestamp->data[3 + offset] - '0') - 1;
-	stamp.tm_mday = (expiry_timestamp->data[4 + offset] - '0') * 10 +
-					(expiry_timestamp->data[5 + offset] - '0');
-	stamp.tm_hour = (expiry_timestamp->data[6 + offset] - '0') * 10 +
-					(expiry_timestamp->data[7 + offset] - '0');
-	stamp.tm_min = (expiry_timestamp->data[8 + offset] - '0') * 10 +
-				   (expiry_timestamp->data[9 + offset] - '0');
-	stamp.tm_sec = (expiry_timestamp->data[10 + offset] - '0') * 10 +
-				   (expiry_timestamp->data[11 + offset] - '0');
-	stamp.tm_isdst = -1;
-
-	time_t tm_t = timegm(&stamp);
-	double time_left = difftime(tm_t, time(NULL));
+	double time_left = difftime(expiry_time, time(NULL));
 	int days_left = (int)(time_left / 86400);
-	char *timeZone = getenv("TZ");
-	setenv("TZ", "GMT", 1);
-	tzset();
-
 	char timestamp[50] = "";
-	strftime(timestamp, 50, "%c %z", localtime(&tm_t));
-	if (timeZone) {
-		setenv("TZ", timeZone, 1);
-	} else {
-		unsetenv("TZ");
-	}
-
-	tzset();
+	np_net_format_timestamp(expiry_time, timestamp, sizeof(timestamp));
 
 	int time_remaining;
 	if (days_left > 0 && days_left <= days_till_exp_warn) {

--- a/plugins/tests/test_sslutils.c
+++ b/plugins/tests/test_sslutils.c
@@ -1,0 +1,302 @@
+/*****************************************************************************
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *
+ *****************************************************************************/
+
+#include "common.h"
+#include "netutils.h"
+#include "../tap/tap.h"
+#include "../../lib/output.h"
+
+#ifdef HAVE_SSL
+#	include <openssl/x509.h>
+#	include <openssl/pem.h>
+#	include <openssl/err.h>
+
+/* prototypes for internal functions*/
+mp_state_enum np_net_ssl_check_certificate(X509 *certificate, int days_till_exp_warn,
+										   int days_till_exp_crit);
+retrieve_expiration_time_result np_net_ssl_get_cert_expiration(X509 *certificate);
+mp_subcheck mp_net_ssl_check_certificate(X509 *certificate, int days_till_exp_warn,
+										 int days_till_exp_crit);
+#endif
+
+const char *progname = "test_sslutils";
+void print_usage(void) {}
+
+#ifdef HAVE_SSL
+static X509 *create_test_cert(long seconds) {
+	X509 *cert = X509_new();
+	if (!cert) {
+		return NULL;
+	}
+
+	ASN1_INTEGER_set(X509_get_serialNumber(cert), 1);
+	X509_set_version(cert, 2);
+
+	X509_gmtime_adj(X509_get_notBefore(cert), 0);
+	X509_gmtime_adj(X509_get_notAfter(cert), seconds);
+
+	X509_NAME *name = X509_get_subject_name(cert);
+	X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"test.example.com", -1,
+							   -1, 0);
+	X509_set_issuer_name(cert, name);
+
+	EVP_PKEY *pkey = NULL;
+	EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+	if (!kctx) {
+		X509_free(cert);
+		return NULL;
+	}
+	if (EVP_PKEY_keygen_init(kctx) <= 0 || EVP_PKEY_CTX_set_rsa_keygen_bits(kctx, 2048) <= 0 ||
+		EVP_PKEY_keygen(kctx, &pkey) <= 0) {
+		EVP_PKEY_CTX_free(kctx);
+		X509_free(cert);
+		return NULL;
+	}
+	EVP_PKEY_CTX_free(kctx);
+
+	X509_set_pubkey(cert, pkey);
+	X509_sign(cert, pkey, EVP_sha256());
+	EVP_PKEY_free(pkey);
+
+	return cert;
+}
+#endif
+
+int main(int argc, char **argv) {
+#ifdef HAVE_SSL
+	plan_tests(37);
+
+	X509 *cert;
+	retrieve_expiration_time_result expiration_time_result;
+	mp_state_enum mp_result;
+	mp_subcheck sc_result;
+#	define TIME_DELTA 5
+
+	/* Valid cert expiring in 30 days */
+	cert = create_test_cert(60 * 60 * 24 * 30);
+	ok(cert != NULL, "Created test cert expiring in 30 days");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_OK, "Cert expiring in 30 days returns STATE_OK (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 10 days */
+	cert = create_test_cert(60 * 60 * 24 * 10);
+	ok(cert != NULL, "Created test cert expiring in 10 days");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_WARNING,
+		   "Cert expiring in 10 days returns STATE_WARNING (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 5 days */
+	cert = create_test_cert(60 * 60 * 24 * 5);
+	ok(cert != NULL, "Created test cert expiring in 5 days");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 5 days returns STATE_CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 1 hour */
+	cert = create_test_cert(60 * 60 * 1);
+	ok(cert != NULL, "Created test cert expiring in 1 hour");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 1 hour returns CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Cert expiring in 30 minutes */
+	cert = create_test_cert(60 * 30);
+	ok(cert != NULL, "Created test cert expiring in 30 minutes");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 30 minutes returns CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Expired cert - should be CRITICAL */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	ok(cert != NULL, "Created test cert that expired 10 days ago");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL, "Expired cert returns STATE_CRITICAL");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - NULL certificate */
+	expiration_time_result = np_net_ssl_get_cert_expiration(NULL);
+	ok(expiration_time_result.errors == NO_SERVER_CERTIFICATE_PRESENT,
+	   "NULL certificate returns NO_SERVER_CERTIFICATE_PRESENT error");
+
+	/* np_net_ssl_get_cert_expiration - valid cert */
+	cert = create_test_cert(60 * 60 * 24 * 30);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK, "Valid cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds > (60 * 60 * 24 * 30) - TIME_DELTA &&
+			   expiration_time_result.remaining_seconds < (60 * 60 * 24 * 30) + TIME_DELTA,
+		   "remaining_seconds is 30 days +- %i", TIME_DELTA);
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - expired cert */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK, "Expired cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds < 0,
+		   "remaining_seconds is negative for expired cert");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	struct tm target_date = {0};
+	target_date.tm_year = 2051 - 1900;
+	target_date.tm_mon = 0;
+	target_date.tm_mday = 1;
+	target_date.tm_hour = 0;
+	target_date.tm_min = 0;
+	target_date.tm_sec = 0;
+	target_date.tm_isdst = -1;
+	time_t time_2051 = mktime(&target_date);
+	time_t time_now = time(NULL);
+	long seconds_remaining = time_2051 - time_now;
+	cert = create_test_cert(seconds_remaining);
+
+	/* GENERALIZEDTIME cert (year > 2049) */
+	ok(cert != NULL, "Created GENERALIZEDTIME cert (2051)");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_OK, "GENERALIZEDTIME cert expiring in 2051 returns STATE_OK");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - GENERALIZEDTIME cert */
+	cert = create_test_cert(seconds_remaining);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK,
+		   "GENERALIZEDTIME cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds > seconds_remaining - TIME_DELTA &&
+			   expiration_time_result.remaining_seconds < seconds_remaining + TIME_DELTA,
+		   "remaining_seconds is correct for Midnight on January 1st 2051 +- %i", TIME_DELTA);
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - NULL certificate */
+	sc_result = mp_net_ssl_check_certificate(NULL, 14, 7);
+	ok(sc_result.state == STATE_CRITICAL, "mp: NULL certificate returns STATE_CRITICAL");
+	ok(sc_result.output != NULL, "mp: NULL certificate sets output");
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 30 days */
+	cert = create_test_cert(60 * 60 * 24 * 30);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_OK,
+		   "mp: Cert expiring in 30 days returns STATE_OK (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 30 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 10 days */
+	cert = create_test_cert(60 * 60 * 24 * 10);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_WARNING,
+		   "mp: Cert expiring in 10 days returns STATE_WARNING (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 10 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 5 days */
+	cert = create_test_cert(60 * 60 * 24 * 5);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 5 days returns STATE_CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 5 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 1 hour */
+	cert = create_test_cert(60 * 60 * 1);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 1 hour returns CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 1 hour sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - cert expiring in 30 minutes */
+	cert = create_test_cert(60 * 30);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 30 minutes returns CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 30 minutes sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - expired cert */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL, "mp: Expired cert returns STATE_CRITICAL");
+		ok(sc_result.output != NULL, "mp: Expired cert sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - GENERALIZEDTIME cert (year > 2049) */
+	cert = create_test_cert(seconds_remaining);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_OK,
+		   "mp: GENERALIZEDTIME cert expiring in 2051 returns STATE_OK");
+		ok(sc_result.output != NULL, "mp: GENERALIZEDTIME cert sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+#else
+	plan_skip_all("SSL support not compiled in");
+#endif
+
+	return exit_status();
+}

--- a/plugins/tests/test_sslutils.c
+++ b/plugins/tests/test_sslutils.c
@@ -27,6 +27,8 @@
 #	include <openssl/err.h>
 
 /* prototypes for internal functions*/
+int np_net_asn1_time_to_time_t(const ASN1_TIME *asn1_time, time_t *out);
+void np_net_format_timestamp(time_t t, char *buf, size_t buflen);
 mp_state_enum np_net_ssl_check_certificate(X509 *certificate, int days_till_exp_warn,
 										   int days_till_exp_crit);
 retrieve_expiration_time_result np_net_ssl_get_cert_expiration(X509 *certificate);
@@ -79,7 +81,7 @@ static X509 *create_test_cert(long seconds) {
 
 int main(int argc, char **argv) {
 #ifdef HAVE_SSL
-	plan_tests(37);
+	plan_tests(44);
 #	define TIME_DELTA 5
 
 	/* Valid cert expiring in 30 days */
@@ -236,6 +238,47 @@ int main(int argc, char **argv) {
 	} else {
 		skip(2, "Cert creation failed");
 	}
+
+	/*np_net_asn1_time_to_time_t with valid UTCTIME */
+	ASN1_TIME *asn1_time = ASN1_TIME_new();
+	ASN1_TIME_set_string(asn1_time, "301128210211Z"); /* Nov 28, 2030 21:02:11 UTC */
+	time_t result;
+	ok(np_net_asn1_time_to_time_t(asn1_time, &result) == 1,
+	   "np_net_asn1_time_to_time_t succeeds with valid UTCTIME");
+	struct tm *tm = gmtime(&result);
+	ok(tm->tm_year == 130 && tm->tm_mon == 10 && tm->tm_mday == 28 && tm->tm_hour == 21 &&
+		   tm->tm_min == 2 && tm->tm_sec == 11,
+	   "UTCTIME correctly converts to Nov 28, 2030 21:02:11 UTC");
+	ASN1_TIME_free(asn1_time);
+
+	/* np_net_asn1_time_to_time_t with valid GENERALIZEDTIME */
+	asn1_time = ASN1_TIME_new();
+	ASN1_TIME_set_string(asn1_time, "20510701120000Z"); /* Jul 1, 2051 12:00:00 UTC */
+	ok(np_net_asn1_time_to_time_t(asn1_time, &result) == 1,
+	   "np_net_asn1_time_to_time_t succeeds with valid GENERALIZEDTIME");
+	tm = gmtime(&result);
+	ok(tm->tm_year == 151 && tm->tm_mon == 6 && tm->tm_mday == 1 && tm->tm_hour == 12 &&
+		   tm->tm_min == 0 && tm->tm_sec == 0,
+	   "GENERALIZEDTIME correctly converts to Jul 1, 2051 12:00:00 UTC");
+	ASN1_TIME_free(asn1_time);
+
+	/* np_net_asn1_time_to_time_t fails with empty ASN1_TIME */
+	asn1_time = ASN1_TIME_new();
+	ok(np_net_asn1_time_to_time_t(asn1_time, &result) == 0,
+	   "np_net_asn1_time_to_time_t fails with empty ASN1_TIME");
+	ASN1_TIME_free(asn1_time);
+
+	/* np_net_format_timestamp produces correct GMT output */
+	asn1_time = ASN1_TIME_new();
+	ASN1_TIME_set_string(asn1_time, "301128210211Z"); /* Nov 28, 2030 21:02:11 UTC */
+	time_t t;
+	np_net_asn1_time_to_time_t(asn1_time, &t);
+	char buf[100] = "";
+	np_net_format_timestamp(t, buf, sizeof(buf));
+	ok(strlen(buf) > 0, "np_net_format_timestamp produces non-empty output");
+	ok(strstr(buf, "+0000") != NULL,
+	   "np_net_format_timestamp output contains +0000 for GMT timezone");
+	ASN1_TIME_free(asn1_time);
 
 	/* mp_net_ssl_check_certificate - valid cert expiring in 5 days */
 	cert = create_test_cert(60 * 60 * 24 * 5);

--- a/plugins/tests/test_sslutils.c
+++ b/plugins/tests/test_sslutils.c
@@ -1,0 +1,298 @@
+/*****************************************************************************
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *
+ *****************************************************************************/
+
+#include "common.h"
+#include "netutils.h"
+#include "../tap/tap.h"
+#include "../../lib/output.h"
+
+#ifdef HAVE_SSL
+#	include <openssl/x509.h>
+#	include <openssl/pem.h>
+#	include <openssl/err.h>
+
+/* prototypes for internal functions*/
+mp_state_enum np_net_ssl_check_certificate(X509 *certificate, int days_till_exp_warn,
+										   int days_till_exp_crit);
+retrieve_expiration_time_result np_net_ssl_get_cert_expiration(X509 *certificate);
+mp_subcheck mp_net_ssl_check_certificate(X509 *certificate, int days_till_exp_warn,
+										 int days_till_exp_crit);
+#endif
+
+const char *progname = "test_sslutils";
+void print_usage(void) {}
+
+#ifdef HAVE_SSL
+static X509 *create_test_cert(long seconds) {
+	X509 *cert = X509_new();
+	if (!cert) {
+		return NULL;
+	}
+
+	ASN1_INTEGER_set(X509_get_serialNumber(cert), 1);
+	X509_set_version(cert, 2);
+
+	X509_gmtime_adj(X509_get_notBefore(cert), 0);
+	X509_gmtime_adj(X509_get_notAfter(cert), seconds);
+
+	X509_NAME *name = X509_get_subject_name(cert);
+	X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"test.example.com", -1,
+							   -1, 0);
+	X509_set_issuer_name(cert, name);
+
+	EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+	if (!kctx) {
+		X509_free(cert);
+		return NULL;
+	}
+	EVP_PKEY *pkey = NULL;
+	if (EVP_PKEY_keygen_init(kctx) <= 0 || EVP_PKEY_CTX_set_rsa_keygen_bits(kctx, 2048) <= 0 ||
+		EVP_PKEY_keygen(kctx, &pkey) <= 0) {
+		EVP_PKEY_CTX_free(kctx);
+		X509_free(cert);
+		return NULL;
+	}
+	EVP_PKEY_CTX_free(kctx);
+
+	X509_set_pubkey(cert, pkey);
+	X509_sign(cert, pkey, EVP_sha256());
+	EVP_PKEY_free(pkey);
+
+	return cert;
+}
+#endif
+
+int main(int argc, char **argv) {
+#ifdef HAVE_SSL
+	plan_tests(37);
+#	define TIME_DELTA 5
+
+	/* Valid cert expiring in 30 days */
+	X509 *cert = create_test_cert(60 * 60 * 24 * 30);
+	ok(cert != NULL, "Created test cert expiring in 30 days");
+	mp_state_enum mp_result;
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_OK, "Cert expiring in 30 days returns STATE_OK (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 10 days */
+	cert = create_test_cert(60 * 60 * 24 * 10);
+	ok(cert != NULL, "Created test cert expiring in 10 days");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_WARNING,
+		   "Cert expiring in 10 days returns STATE_WARNING (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 5 days */
+	cert = create_test_cert(60 * 60 * 24 * 5);
+	ok(cert != NULL, "Created test cert expiring in 5 days");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 5 days returns STATE_CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Valid cert expiring in 1 hour */
+	cert = create_test_cert(60 * 60 * 1);
+	ok(cert != NULL, "Created test cert expiring in 1 hour");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 1 hour returns CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Cert expiring in 30 minutes */
+	cert = create_test_cert(60 * 30);
+	ok(cert != NULL, "Created test cert expiring in 30 minutes");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL,
+		   "Cert expiring in 30 minutes returns CRITICAL (threshold 14/7)");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* Expired cert - should be CRITICAL */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	ok(cert != NULL, "Created test cert that expired 10 days ago");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_CRITICAL, "Expired cert returns STATE_CRITICAL");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - NULL certificate */
+	retrieve_expiration_time_result expiration_time_result = np_net_ssl_get_cert_expiration(NULL);
+	ok(expiration_time_result.errors == NO_SERVER_CERTIFICATE_PRESENT,
+	   "NULL certificate returns NO_SERVER_CERTIFICATE_PRESENT error");
+
+	/* np_net_ssl_get_cert_expiration - valid cert */
+	cert = create_test_cert(60 * 60 * 24 * 30);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK, "Valid cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds > (60 * 60 * 24 * 30) - TIME_DELTA &&
+			   expiration_time_result.remaining_seconds < (60 * 60 * 24 * 30) + TIME_DELTA,
+		   "remaining_seconds is 30 days +- %i", TIME_DELTA);
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - expired cert */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK, "Expired cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds < 0,
+		   "remaining_seconds is negative for expired cert");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	struct tm target_date = {0};
+	target_date.tm_year = 2051 - 1900;
+	target_date.tm_mon = 0;
+	target_date.tm_mday = 1;
+	target_date.tm_hour = 0;
+	target_date.tm_min = 0;
+	target_date.tm_sec = 0;
+	target_date.tm_isdst = -1;
+	time_t time_2051 = mktime(&target_date);
+	time_t time_now = time(NULL);
+	long seconds_remaining = time_2051 - time_now;
+	cert = create_test_cert(seconds_remaining);
+
+	/* GENERALIZEDTIME cert (year > 2049) */
+	ok(cert != NULL, "Created GENERALIZEDTIME cert (2051)");
+	if (cert) {
+		mp_result = np_net_ssl_check_certificate(cert, 14, 7);
+		ok(mp_result == STATE_OK, "GENERALIZEDTIME cert expiring in 2051 returns STATE_OK");
+	} else {
+		skip(1, "Cert creation failed");
+	}
+
+	/* np_net_ssl_get_cert_expiration - GENERALIZEDTIME cert */
+	cert = create_test_cert(seconds_remaining);
+	if (cert) {
+		expiration_time_result = np_net_ssl_get_cert_expiration(cert);
+		ok(expiration_time_result.errors == ALL_OK,
+		   "GENERALIZEDTIME cert returns ALL_OK error code");
+		ok(expiration_time_result.remaining_seconds > seconds_remaining - TIME_DELTA &&
+			   expiration_time_result.remaining_seconds < seconds_remaining + TIME_DELTA,
+		   "remaining_seconds is correct for Midnight on January 1st 2051 +- %i", TIME_DELTA);
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - NULL certificate */
+	mp_subcheck sc_result = mp_net_ssl_check_certificate(NULL, 14, 7);
+	ok(sc_result.state == STATE_CRITICAL, "mp: NULL certificate returns STATE_CRITICAL");
+	ok(sc_result.output != NULL, "mp: NULL certificate sets output");
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 30 days */
+	cert = create_test_cert(60 * 60 * 24 * 30);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_OK,
+		   "mp: Cert expiring in 30 days returns STATE_OK (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 30 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 10 days */
+	cert = create_test_cert(60 * 60 * 24 * 10);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_WARNING,
+		   "mp: Cert expiring in 10 days returns STATE_WARNING (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 10 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 5 days */
+	cert = create_test_cert(60 * 60 * 24 * 5);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 5 days returns STATE_CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 5 days sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - valid cert expiring in 1 hour */
+	cert = create_test_cert(60 * 60 * 1);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 1 hour returns CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 1 hour sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - cert expiring in 30 minutes */
+	cert = create_test_cert(60 * 30);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL,
+		   "mp: Cert expiring in 30 minutes returns CRITICAL (threshold 14/7)");
+		ok(sc_result.output != NULL, "mp: Cert expiring in 30 minutes sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - expired cert */
+	cert = create_test_cert(60 * 60 * 24 * -10);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_CRITICAL, "mp: Expired cert returns STATE_CRITICAL");
+		ok(sc_result.output != NULL, "mp: Expired cert sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+
+	/* mp_net_ssl_check_certificate - GENERALIZEDTIME cert (year > 2049) */
+	cert = create_test_cert(seconds_remaining);
+	if (cert) {
+		sc_result = mp_net_ssl_check_certificate(cert, 14, 7);
+		ok(sc_result.state == STATE_OK,
+		   "mp: GENERALIZEDTIME cert expiring in 2051 returns STATE_OK");
+		ok(sc_result.output != NULL, "mp: GENERALIZEDTIME cert sets output");
+	} else {
+		skip(2, "Cert creation failed");
+	}
+#else
+	plan_skip_all("SSL support not compiled in");
+#endif
+
+	return exit_status();
+}

--- a/plugins/tests/test_sslutils.t
+++ b/plugins/tests/test_sslutils.t
@@ -1,0 +1,6 @@
+#!/usr/bin/perl
+use Test::More;
+if (! -e "./test_sslutils") {
+	plan skip_all => "./test_sslutils not compiled - please enable libtap library to test";
+}
+exec "./test_sslutils";

